java/dagger/lint/DaggerKotlinIssuesDetector.kt

```diff
@@ -0,0 +1,238 @@
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.UMethod
+import org.jetbrains.uast.getUastParentOfType
+import org.jetbrains.uast.kotlin.KotlinUClass
+import org.jetbrains.uast.toUElementOfType
+import dagger.lint.DaggerKotlinIssuesDetector.Companion.ISSUE_FIELD_SITE_TARGET_ON_QUALIFIER_ANNOTATION
+import dagger.lint.DaggerKotlinIssuesDetector.Companion.ISSUE_JVM_STATIC_PROVIDES_IN_OBJECT
+import dagger.lint.DaggerKotlinIssuesDetector.Companion.ISSUE_MODULE_COMPANION_OBJECTS
+import dagger.lint.DaggerKotlinIssuesDetector.Companion.ISSUE_MODULE_COMPANION_OBJECTS_NOT_IN_MODULE_PARENT
+import java.util.EnumSet
+
+/**
+ * This is a simple lint check to catch common Dagger+Kotlin usage issues.
+ *
+ * - [ISSUE_FIELD_SITE_TARGET_ON_QUALIFIER_ANNOTATION] covers using `field:` site targets for member injections, which
+ * are redundant as of Dagger 2.25.
+ * - [ISSUE_JVM_STATIC_PROVIDES_IN_OBJECT] covers using `@JvmStatic` for object `@Provides`-annotated functions, which
+ * are redundant as of Dagger 2.25. @JvmStatic on companion object functions are redundant as of Dagger 2.26.
+ * - [ISSUE_MODULE_COMPANION_OBJECTS] covers annotating companion objects with `@Module`, as they are now part
+ * of the enclosing module class's API in Dagger 2.26. This will also error if the enclosing class is _not_ in a
+ * `@Module`-annotated class, as this object just should be moved to a top-level object to avoid confusion.
+ * - [ISSUE_MODULE_COMPANION_OBJECTS_NOT_IN_MODULE_PARENT] covers annotating companion objects with `@Module` when
+ * the parent class is _not_ also annotated with `@Module`. While technically legal, these should be moved up to
+ * top-level objects to avoid confusion.
+ */
+@Suppress("UnstableApiUsage")
+class DaggerKotlinIssuesDetector : Detector(), SourceCodeScanner {
+
+  companion object {
+    // We use the overloaded constructor that takes a varargs of `Scope` as the last param.
+    // This is to enable on-the-fly IDE checks. We are telling lint to run on both
+    // JAVA and TEST_SOURCES in the `scope` parameter but by providing the `analysisScopes`
+    // params, we're indicating that this check can run on either JAVA or TEST_SOURCES and
+    // doesn't require both of them together.
+    // From discussion on lint-dev https://groups.google.com/d/msg/lint-dev/ULQMzW1ZlP0/1dG4Vj3-AQAJ
+    // This was supposed to be fixed in AS 3.4 but still required as recently as 3.6-alpha10.
+    private val SCOPES = Implementation(DaggerKotlinIssuesDetector::class.java,
+        EnumSet.of(Scope.JAVA_FILE, Scope.TEST_SOURCES),
+        EnumSet.of(Scope.JAVA_FILE),
+        EnumSet.of(Scope.TEST_SOURCES)
+    )
+
+    private val ISSUE_JVM_STATIC_PROVIDES_IN_OBJECT: Issue = Issue.create(
+        "JvmStaticProvidesInObjectDetector",
+        "@JvmStatic used for @Provides function in an object class",
+        """
+          As of Dagger 2.25, it's redundant to annotate @Provides functions in object classes with @JvmStatic.
```

Comment:
"As of Dagger 2.xx" seems fine for an app developed rule but coming from the library I think we can drop it, since the rules are versioned.

Reply:
a307266